apis/python/src/tiledbsc/soma_collection.py

```diff
@@ -125,3 +134,124 @@ def cell_count(self) -> int:
         Returns sum of `soma.cell_count()` over SOMAs in the collection.
         """
         return sum(soma.cell_count() for soma in self)
+
+    # ----------------------------------------------------------------
+    def attribute_filter(
+        self,
+        obs_attr_names: List[str] = [],
+        obs_query_string: str = None,
+        var_attr_names: List[str] = [],
+        var_query_string: str = None,
+    ) -> Optional[SOMASlice]:
+        """
+        Subselects the obs, var, and X/data using the specified queries on obs and var,
+        concatenating across SOMAs in the collection.  Queries use the TileDB-Py `QueryCondition`
+        API. If `obs_query_string` is `None`, the `obs` dimension is not filtered and all of `obs`
+        is used; similiarly for `var`. Return value of `None` indicates an empty slice.
+        """
+
+        soma_slices = []
+        for soma in self:
+            # E.g. querying for 'cell_type == "blood"' but this SOMA doesn't have a cell_type column in
+            # its obs at all.
+            if not soma.obs.has_attr_names(obs_attr_names):
+                continue
+            # E.g. querying for 'feature_name == "MT-CO3"' but this SOMA doesn't have a feature_name
+            # column in its var at all.
+            if not soma.var.has_attr_names(var_attr_names):
+                continue
+
+            soma_slice = soma.attribute_filter(obs_query_string, var_query_string)
+            if soma_slice != None:
+                # print("Slice SOMA from", soma.name, soma.X.data.shape(), "to", soma_slice.ann.X.shape)
+                soma_slices.append(soma_slice)
+
+        return SOMASlice.concat(soma_slices)
```

Comment:
concat() makes assumptions about how the user wants to process the data.  I think it is preferred if you can simply return a list of slices.

that steps around issues such as duplicate IDs.

It might also make it easier to return partially evaluated queries which allow the user to later ask for sub-objects (eg, X, obsm, ..)